cmd/ferretdb/main.go

```diff
@@ -217,12 +221,108 @@ func main() {
 	case "run":
 		run()
 	case "ping":
-		// TODO https://github.com/FerretDB/FerretDB/issues/4246
+		ping()
 	default:
 		panic("unknown sub-command")
 	}
 }
 
+// ping creates connection to FerretDB instance specified by the flags, and runs `ping` command against it.
+// The check is only executed if --setup-database flag is set.
+func ping() {
+	logger := setupLogger(cli.Log.Format, "")
+	checkFlags(logger)
+
+	l := logger.Sugar()
+
+	if cli.Setup.Database == "" {
+		l.Info("Setup database not specified - skipping ping.")
+		return
+	}
+
+	var pingExecuted bool
+
+	if cli.Listen.Addr != "" {
+		pingExecuted = true
+
+		host, port, err := net.SplitHostPort(cli.Listen.Addr)
+		if err != nil {
+			l.Fatal(err)
+		}
+
+		l.Infof("--listen-addr flag is set. Pinging %s...", cli.Listen.Addr)
+
+		if host == "" {
+			host = "127.0.0.1"
+
+			l.Debugf("Host not specified, defaulting to %s.", host)
+		}
+
+		u := &url.URL{
+			Scheme: "mongodb",
+			Host:   fmt.Sprintf("%s:%s", host, port),
+			Path:   cli.Setup.Database,
+			User:   url.UserPassword(cli.Setup.Username, cli.Setup.Password),
+		}
+
+		l.Debugf("Pinging %s...", u)
+
+		ctx, cancel := context.WithTimeout(context.Background(), cli.Setup.Timeout)
+		defer cancel()
+
+		client, err := mongo.Connect(ctx, options.Client().ApplyURI(u.String()))
+		if err != nil {
+			l.Fatal(err)
+		}
+
+		defer func() {
+			if err = client.Disconnect(ctx); err != nil {
+				l.Fatal(err)
+			}
+		}()
+
+		if err = client.Ping(ctx, nil); err != nil {
+			l.Fatal(err)
+		}
+
+		l.Info("Ping successful.")
+	}
+
+	if cli.Listen.Unix != "" {
+		pingExecuted = true
+
+		ctx, cancel := context.WithTimeout(context.Background(), cli.Setup.Timeout)
+		defer cancel()
+
+		l.Infof("--listen-unix flag is set. Pinging %s...", cli.Listen.Unix)
+
+		u := "mongodb://" + url.PathEscape(cli.Listen.Unix)
+
+		l.Debugf("Pinging %s...", u)
+
+		client, err := mongo.Connect(ctx, options.Client().ApplyURI(u))
+		if err != nil {
+			l.Fatal(err)
+		}
+
+		defer func() {
+			if err = client.Disconnect(ctx); err != nil {
+				l.Fatal(err)
+			}
+		}()
+
+		if err = client.Ping(ctx, nil); err != nil {
+			l.Fatal(err)
+		}
+
+		l.Info("Ping successful.")
```

Comment:
That's the same code as in another branch, we should refactor it: create URI first, then use it